OxfordInstruments Proteox driver

abioxinst opened the pull request:
-	In the `driver` directory, we have changed the directory name from `Oxford` to `OxfordInstruments`. 
-	Within `OxfordInstruments`, you will now find our QCoDeS driver, `Proteox.py`, and a `README.md` file to explain how to use it. 
-	To use `Proteox.py`, our decs-visa application is required. Therefore, this has been added as a submodule within the `OxfordInstruments` directory. 
-	Within the `docs/examples` directory, you will find an example Jupyter Notebook, titled `OxfordInstruments_Proteox.ipynb`.

abioxinst commented on src/qcodes_contrib_drivers/drivers/OxfordInstruments/img/DECS_VISA.jpg:
The images in the src folder have been removed. All images are now in the example folder.

abioxinst commented on docs/examples/OxfordInstruments_Proteox.ipynb:
```diff
@@ -0,0 +1,402 @@
+    "import qcodes as qc\n",
+    "from qcodes.logger.logger import start_all_logging\n",
+    "\n",
+    "sys.path.append(\"../../src/qcodes_contrib_drivers/drivers/OxfordInstruments/\")\n",
```
The file has been updated to replace paths with imports for qcodes_contrib_driver.

abioxinst commented on src/qcodes_contrib_drivers/drivers/OxfordInstruments/README.md:
```diff
@@ -0,0 +1,97 @@
+    from src.decs_visa_tools.decs_visa_settings import PORT
+    from src.decs_visa_tools.decs_visa_settings import HOST
+    from src.decs_visa_tools.decs_visa_settings import SHUTDOWN
+    from src.decs_visa_tools.decs_visa_settings import WRITE_DELIM
```
The file has been updated to replace paths with imports for qcodes_contrib_driver.

abioxinst commented on .gitmodules:
```diff
@@ -0,0 +1,3 @@
+[submodule "src/qcodes_contrib_drivers/drivers/OxfordInstruments/decsvisa"]
+	path = src/qcodes_contrib_drivers/drivers/OxfordInstruments/decsvisa
+	url = git@github.com:decs-visa/decsvisa.git
```
We would like to keep decsvisa as a submodule for the meantime, as the idea is users are able to add additional commands in the command_dictionary. We have renamed this directory to _decsvisa to make it clear that this is not public API as you suggested.

abioxinst replied:
Sorry, my mistake! I was pushing to decs-visa/Qcodes_contrib_drivers and not abioxinst/Qcodes_contrib_driver. I have now merged the changes across to abioxinst:main so you should now see these updates.

abioxinst commented on src/qcodes_contrib_drivers/drivers/OxfordInstruments/README.md:
`````diff
@@ -0,0 +1,97 @@
+````python
+if running_on.startswith("Windows"):
+    print(f"Running on {running_on} - start subprocess without PIPEd output")
+    subprocess.Popen(["python", decs_visa_path])
`````
decsvisa can be used independently to QCoDeS, which requires the user modifying some settings in decsvisa. 

We are happy to look into how we might be able to automatically start this server for QCoDeS users, but still keep it flexible for other users. 

For the meantime, starting the server with the subprocess package seems to work well for accommodating both use cases.

jenshnielsen replied:
ok this is fine for now

jenshnielsen commented:
Some of the current `run mypy and pytest` failiurs should be fixed by https://github.com/QCoDeS/Qcodes_contrib_drivers/pull/360

jenshnielsen commented:
It looks like the the `_decsvisa` currently has some problems type checking with mypy. If this is not something that you want to fix right now you can exclude it from type checking by adding it to this list in pyproject.toml

```toml
[[tool.mypy.overrides]]
module = [
    "qcodes_contrib_drivers.drivers.Spectrum.pyspcm"
    ]
ignore_errors = true

```

jenshnielsen commented on src/qcodes_contrib_drivers/drivers/OxfordInstruments/Proteox.py:
```diff
@@ -0,0 +1,583 @@
+from qcodes.instrument import VisaInstrument
+from qcodes.parameters import MultiParameter
+
+from qcodes_contrib_drivers.drivers.OxfordInstruments.Proteox._decsvisa.src.decs_visa_tools.decs_visa_settings import PORT
```
I think the docs build fail because these path are wrong. As far as I can see they should not contain Proteox
e.g. they should be.

```suggestion
from qcodes_contrib_drivers.drivers.OxfordInstruments._decsvisa.src.decs_visa_tools.decs_visa_settings import PORT
```

jenshnielsen replied:
and similar for the 3 lines below

abioxinst replied:
This has been corrected.

jenshnielsen commented on src/qcodes_contrib_drivers/drivers/OxfordInstruments/README.md:
`````diff
@@ -0,0 +1,90 @@
+5.  Import the `QCoDeS` driver and connect to `oi.DECS` as follows:
+
+    ````python
+    from Proteox import oiDECS
`````
Is this importing the driver from this repo? In that case the import should normally be something like `qcodes_contrib_drivers.drivers.Oxford.Proteox import ...`

abioxinst replied:
This has been corrected to your suggestion.

jenshnielsen commented:
@abioxinst could you click the Update branch button (or manually merge/rebase main) that should fix some errors from github actions that have been fixed on main. Once that is done I left a few more inline comments (mostly optional) the main thing to merge this is that we need to make github actions go grean

abioxinst commented on src/qcodes_contrib_drivers/drivers/OxfordInstruments/README.md:
```diff
@@ -0,0 +1,90 @@
+    DUAL_TURBO_FITTED=False # not currently used
+
+    # Does the system have a 3He flow meter
+    HE3_FLOW_METER_FITTED=True
```
For now we are happy to leave the configuration set up as it is. We are looking into how to do this better in our ongoing work on drivers for our cryostats.

abioxinst commented:
> It looks like the the `_decsvisa` currently has some problems type checking with mypy. If this is not something that you want to fix right now you can exclude it from type checking by adding it to this list in pyproject.toml
> 
> ```toml
> [[tool.mypy.overrides]]
> module = [
>     "qcodes_contrib_drivers.drivers.Spectrum.pyspcm"
>     ]
> ignore_errors = true
> ```

We are addressing some of these issues. Currently in the process of making improvements to the code so it will pass this test.

jenshnielsen commented on pyproject.toml:
```diff
@@ -66,6 +66,7 @@ show_error_codes = true
 [[tool.mypy.overrides]]
 module = [
     "qcodes_contrib_drivers.drivers.Spectrum.pyspcm"
```
```suggestion
    "qcodes_contrib_drivers.drivers.Spectrum.pyspcm",
```

abioxinst commented:
> > It looks like the the `_decsvisa` currently has some problems type checking with mypy. If this is not something that you want to fix right now you can exclude it from type checking by adding it to this list in pyproject.toml
> > ```toml
> > [[tool.mypy.overrides]]
> > module = [
> >     "qcodes_contrib_drivers.drivers.Spectrum.pyspcm"
> >     ]
> > ignore_errors = true
> > ```
> 
> We are addressing some of these issues. Currently in the process of making improvements to the code so it will pass this test.

We have now addressed the issues and _decsvisa should now pass a type checking with mypy, providing the flag `--disable-error-code=import-untyped` is used to suppress all warning from imported libraries.

jenshnielsen commented:
> We have now addressed the issues and _decsvisa should now pass a type checking with mypy, providing the flag --disable-error-code=import-untyped is used to suppress all warnings from imported libraries.

This is not a great solution since that would trigger everywhere for all untyped modules. In CI we were still seeing errors from descvisa so I have disabled it for now. 

I think the reason is partially that descvisa has dependencies on autobahm and dotenv but since you are not actually installing descvisa you will not have these installed. I think you need to add python-dotenv and autobahn to the dependencies of this package in pyproject.toml
